rpcs3/Emu/RSX/Common/ProgramStateCache.h

```diff
@@ -195,51 +199,77 @@ class program_state_cache
 	/// bool here to inform that the program was preexisting.
 	std::tuple<const vertex_program_type&, bool> search_vertex_program(const RSXVertexProgram& rsx_vp, bool force_load = true)
 	{
-		const auto& I = m_vertex_shader_cache.find(rsx_vp);
-		if (I != m_vertex_shader_cache.end())
+		bool recompile = false;
+		vertex_program_type* new_shader;
 		{
-			return std::forward_as_tuple(I->second, true);
-		}
+			reader_lock lock(m_vertex_mutex);
 
-		if (!force_load)
-		{
-			return std::forward_as_tuple(__null_vertex_program, false);
+			const auto& I = m_vertex_shader_cache.find(rsx_vp);
+			if (I != m_vertex_shader_cache.end())
+			{
+				return std::forward_as_tuple(I->second, true);
+			}
+
+			if (!force_load)
+			{
+				return std::forward_as_tuple(__null_vertex_program, false);
+			}
+
+			LOG_NOTICE(RSX, "VP not found in buffer!");
+
+			lock.upgrade();
+			auto [it, inserted] = m_vertex_shader_cache.try_emplace(rsx_vp);
+			new_shader = &(it->second);
+			recompile = inserted;
 		}
 
-		LOG_NOTICE(RSX, "VP not found in buffer!");
-		vertex_program_type& new_shader = m_vertex_shader_cache[rsx_vp];
-		backend_traits::recompile_vertex_program(rsx_vp, new_shader, m_next_id++);
+		if(recompile)
```

Comment:
Formatting nitpick, formatting rules require both a space after 'if' and that conditions be braced. You can see the force_load condition above, follow that layout.